### PR TITLE
Adds llvm-cov tool

### PR DESCRIPTION
**Pull request by RubenRBS**
Moves base-tool to ts, but this is a min effort one

Fixes an issue with tools.js pane.
onCompileResult did not reset the plainContentRoot for toolResults, if no languageId is defined for the current function

### CODE REVIEW

File: lib/tooling/base-tool.ts [hunk 1]
```diff
@@ -30,18 +30,29 @@ import _ from 'underscore';
 import * as exec from '../exec';
 import {logger} from '../logger';
 import * as utils from '../utils';
+import {ToolEnv, ToolInfo} from './base-tool.interface';
```
Comment (RubenRBS) [inline lib/tooling/base-tool.ts:33]: This was the min needed to get it working. I'll properly do this before merging the PR

File: lib/tooling/llvm-cov-tool.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                    '-instr-profile=./coverage.profdata',
+                    '-format',
+                    'text',
+                    '-use-color',
+                    '-Xdemangler',
+                    cppFiltPath,
+                    '-Xdemangler',
+                    '-n',
+                    '-compilation-dir=./',
+                    ...args,
```
Comment (RubenRBS) [inline lib/tooling/llvm-cov-tool.ts:91]: The args are meant to be any/all of `-show-instantiations=true -show-expansions=true -show-line-counts -show-branches=percent -show-regions`

File: static/panes/tool.js [hunk 3]
```diff
@@ -491,6 +491,7 @@ Tool.prototype.onCompileResult = function (id, compiler, result) {
             if (toolResult.languageId) {
                 this.setEditorContent(_.pluck(toolResult.stdout, 'text').join('\n'));
             } else {
+                this.plainContentRoot.empty();
```
Comment (RubenRBS) [inline static/panes/tool.js:494]: This is the bug referenced in the PR message

File: lib/tooling/llvm-cov-tool.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    override async runTool(compilationInfo, inputFilepath, args, stdin /*, supportedLibraries*/) {
+        const compilationExecOptions = this.getDefaultExecOptions() as ExecutionOptions;
+        compilationExecOptions.customCwd = path.dirname(inputFilepath);
+        compilationExecOptions.input = stdin;
+        try {
+            const compilationResult = await this.exec(
+                compilationInfo.compiler.exe,
+                ['-fprofile-instr-generate', '-fcoverage-mapping', '-g', '-O0', inputFilepath, '-o', 'coverage.a'],
```
Comment (RubenRBS) [inline lib/tooling/llvm-cov-tool.ts:44]: Got this from Hana's comment in Discord - Let me know if I should change anything

File: lib/tooling/llvm-cov-tool.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                    `<Compilation error>\n${compilationResult.stdout}\n${compilationResult.stderr}`,
+                );
+            }
+
+            const runExecOptions = this.getDefaultExecOptions() as ExecutionOptions;
+            runExecOptions.customCwd = path.dirname(inputFilepath);
+
+            await this.exec('./coverage.a', [], {
+                ...runExecOptions,
+                input: stdin,
```
Comment (RubenRBS) [inline lib/tooling/llvm-cov-tool.ts:59]: Thought that being able to set stdin to the user program would be useful

**Review by RubenRBS — COMMENTED** [review]
Some thoughts

File: lib/tooling/base-tool.interface.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    monacoStdin?: string;
+    icon?: string;
+    darkIcon?: string;
+    compilerLanguage: LanguageKey;
+};
+
+export type ToolEnv = {
+    ceProps: (key: string, defaultValue?: any) => string | boolean | number | undefined;
+    compilerProps: (key: string, defaultValue?: any) => string | boolean | number | undefined;
+};
```
Comment (RubenRBS) [inline lib/tooling/base-tool.interface.ts:49]: This is the only that needs work now. The compilerProps type belongs in base-compiler

File: lib/tooling/base-tool.ts [hunk 7]
```diff
@@ -27,9 +27,15 @@ import path from 'path';
...
 import _ from 'underscore';
 
+import {ExecutionOptions, ToolResult} from '../../types/compilation/compilation.interfaces';
+import {ResultLine} from '../../types/resultline/resultline.interfaces';
 import * as exec from '../exec';
 import {logger} from '../logger';
-import * as utils from '../utils';
+import {parseOutput} from '../utils';
+
+import {ToolEnv, ToolInfo, ToolTypeKey} from './base-tool.interface';
```
Comment (RubenRBS) [inline lib/tooling/base-tool.ts:36]: Why does not `make check` show this?? We should fix that

File: lib/tooling/base-tool.ts [hunk 8]
```diff
@@ -38,19 +44,21 @@ const toolCounter = new PromClient.Counter({
...
+    protected tool: ToolInfo;
+    private env: ToolEnv;
+    private addOptionsToToolArgs = true;
+
+    constructor(toolInfo: ToolInfo, env: ToolEnv) {
         this.tool = toolInfo;
         this.env = env;
-        this.tool.exclude = this.tool.exclude ? this.tool.exclude.split(':') : [];
         this.addOptionsToToolArgs = true;
-        this.parseOutput = utils.parseOutput;
```
Comment (RubenRBS) [inline lib/tooling/base-tool.ts:46]: parseOutput is now a protected member function, which can be overriden by subclasses if needed

File: lib/tooling/base-tool.ts [hunk 9]
```diff
@@ -60,81 +68,88 @@ export class BaseTool {
...
             languageId: 'stderr',
             stdout: [],
-            stderr: utils.parseOutput(message),
+            stderr: this.parseOutput(message),
         };
     }
 
     // mostly copy&paste from base-compiler.js
-    findLibVersion(selectedLib, supportedLibraries) {
+    findLibVersion(selectedLib: SelectedLibraryVersion, supportedLibraries: Record<string, Library>) {
```
Comment (RubenRBS) [inline lib/tooling/base-tool.ts:111]: Same modifications as in base-compiler

File: lib/tooling/base-tool.ts [hunk 10]
```diff
@@ -60,81 +68,88 @@ export class BaseTool {
...
-                return foundVersion.options;
-            }),
-        );
+            return foundVersion.options;
+        });
     }
 
-    async runTool(compilationInfo, inputFilepath, args, stdin /*, supportedLibraries*/) {
+    async runTool(
+        compilationInfo: Record<any, any>,
```
Comment (RubenRBS) [inline lib/tooling/base-tool.ts:140]: Needs a more specific type, but I wasn't sure which

**Review by RubenRBS — COMMENTED** [review]
Some more comments

File: lib/tooling/base-tool.ts (hunk 10 above)
Comment (partouf) [reply lib/tooling/base-tool.ts:140]: This is whatever getCompilationInfo returns

Comment (RubenRBS) [reply lib/tooling/base-tool.ts:140]: yes, but as that function extends one of its parameters, and that one does not have type info either, I'm not too sure what the best thing to do is

**Review by partouf — APPROVED** [review]

**Comment by RubenRBS** [comment]
Just in case, testing with all live tools in staging before merging

**Pull request merged**